### ISSUE TITLE
Lines are ignored on Darwin but not on Linux

### ISSUE THREAD

**Issue by RossComputerGuy** [mesonbuild/meson#10932]
**Describe the bug**
I have [this `meson.build`](https://github.com/ExpidusOS/genesis/blob/0efb900b9a26bce45b4bb15cdfac3b98dc0468b2/plugins/gtk3/meson.build) file which is used for adding a plugin to my program. In the parent directory, there is a file which is used to include this meson file if the dependencies are found. I have discovered that despite the dependencies being found, the `plugin-gtk3-prettier` target being added, and the files are configured, the `shared_module()` does **not** compile. It has no target, I've even discovered that I can change it to `executable()` and all the `library()` variations but it does not compile. I can change it to a non-existent function name like `sdjfjsdildsfewf` and no configure error. If I put bad syntax in it (eg. `;';';';';''p[]\;';][`) and it does throw a configure error.

However, despite all of this, the lines do compile on Linux but not on Darwin even though both are on `v0.61.2`.

**To Reproduce**

Unsure how to reproduce.

**Expected behavior**

Meson should actually not be ignoring these lines and should be doing stuff beyond parsing on lines 25 to 32.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)?: Native
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.): macOS Monterey 12.5.1
* what Python version are you using e.g. 3.8.0: 3.10.6
* what `meson --version`: 0.61.2
* what `ninja --version` if it's a Ninja build: 1.10.2

System is using the `nix` package manager for dependencies.

Edit: tried with `brew` and v0.63.2 and v0.63.3 are affected. I managed to [fix it with this commit](https://github.com/ExpidusOS/genesis/commit/087c33ea38dd3956d9973f6245c0c7ab675ba962) and I am unsure how that fixes it.

**Comment by tristan957** [comment]
Have you tried adding `build_by_default: true`? Can't remember if that is the default or not. I don't see any targets of yours which depend on the plugins array

**Comment by RossComputerGuy** [comment]
No but the docs said its default to `true` and that wouldn't explain why it would build on Linux but not show up on Darwin. The weirdest thing is I could put in arguments that don't exist and I would not get any errors.

**Comment by tristan957** [comment]
> I could put in arguments that don't exist and I would not get any errors.

Could you give an example?

What backend are you using in Darwin? Ninja, xcode?

**Comment by RossComputerGuy** [comment]
Ninja, I could add in an argument like this `fuck_fuck_fuck_fuck: 1` and it wouldn't say anything besides that it reconfigured and 0 targets were changed.

**Comment by tristan957** [comment]
@dcbaker I assume this is a bug that would get fixed with your build target stuff?

**Comment by tristan957** [comment]
If you could pin down the original issue that would be great, there aren't many Darwin people among the regular contributors

**Comment by eli-schwartz** [comment]
When you say "lines are ignored"...

In `plugins/meson.build`, you have:
```meson
subdir('gtk3', if_found: gtk3)
```

What happens if you add, immediately above that, `message('found gtk?', gtk3.found())`?

In `plugins/gtk3/meson.build`, do you get a print message if the first line is `message('I have reached plugins/gtk3/meson.build')`?

You said that invalid syntax does trigger an error on Linux, but not on Darwin?

**Comment by RossComputerGuy** [comment]
GTK is found, `configure_file` commands are executed but not the `shared_module` or any binary target command. Message does get printed that is put into `plugins/gtk3/meson.build`. If I put in invalid syntax, both will show that. The only problem is the binary targets aren't being added in on Darwin.

**Comment by RossComputerGuy** [comment]
```diff
diff --git a/plugins/gtk3/meson.build b/plugins/gtk3/meson.build
index aa19f63..5261c39 100644
--- a/plugins/gtk3/meson.build
+++ b/plugins/gtk3/meson.build
@@ -22,13 +22,12 @@ endif

 plugin_gtk3_sources += [build_conf_plugin_gtk3]

-plugin_gtk3 = shared_module('genesis-shell-gtk3', build_conf_plugin_gtk3, plugin_gtk3_rel_sources,
+plugins += [shared_module('genesis-shell-gtk3', build_conf_plugin_gtk3, plugin_gtk3_rel_sources,
   c_args: ['-DG_LOG_DOMAIN="GenesisShellGtk3"', '-DGETTEXT_PACKAGE="genesis-shell"'],
   vala_args: ['--define=GETTEXT_PACKAGE="genesis-shell"'],
   vala_gir: 'GenesisShellGtk3-@0@.gir'.format(shortver.split('-')[0]),
-  dependencies: [libgenesis_core, gtk3, gtk_layer_shell],
+  name_suffix: 'so',
+  dependencies: [libgenesis_core, gtk3],
   install: true,
   install_dir: [join_paths(libdir, 'genesis-shell', 'plugins'), join_paths(datadir, 'include', 'genesis-shell', 'plugins'),
-    join_paths(datadir, 'genesis-shell', 'plugins', 'vapi'), join_paths(datadir, 'genesis-shell', 'plugins', 'gir-1.0')])
-
-plugins += [plugin_gtk3]
+    join_paths(datadir, 'genesis-shell', 'plugins', 'vapi'), join_paths(datadir, 'genesis-shell', 'plugins', 'gir-1.0')])]
```
This diff shows the changes I had to make to the file to get it to work on Darwin. The changes shouldn't affect the way Meson handles the file but it does.

**Comment by eli-schwartz** [comment]
I would therefore usually ask you to add a message() before and after that code, to prove it's still getting parsed. But I think the issue here is evident.

Most of those changes are completely meaningless. There's one meaningful change, though, which is that you removed the dependency on gtk_layer_shell.

This tells me that gtk_layer_shell is a disabler object, which means the interpreter goes through the following flow:
- parse and lex the file, line by line, which does mean choking on invalid syntax that cannot be parsed
- assemble an AST which looks something like `PlusAssignmentNode(Literal('plugins'), List(FunctionCall('shared_module', bunch_of_args)))`
  - bunch_of_args includes a disabler object, which disables parsing of FunctionCall, and similarly propagating upward
- since FunctionCall is never actually executed as code (actually, it *is* parsed as code: the code it parses is `FunctionCall('do_nothing_and_ignore_args', bunch_of_args)`), we never bother looking it up in the function namespace, verifying its arguments, or producing build rules for it.
- If you have additional lines after this, they should be executed as code, above and beyond merely being parsed for syntax well-formedness.


Basically, Meson is behaving as expected. But perhaps you didn't realize this was what a disabler actually does. https://mesonbuild.com/Reference-manual_returned_disabler.html

**Comment by RossComputerGuy** [comment]
Ah ok, that makes more sense. And I was adding `message()` calls around locally.
